aten/src/ATen/native/cuda/ConvolutionMM2d.cu

```diff
@@ -1,12 +1,525 @@
 #include <ATen/ATen.h>
-#include <ATen/LegacyTHFunctionsCUDA.h>
 
-namespace at {
-namespace native {
+#include <ATen/div_rtn.h>
+#include <ATen/AccumulateType.h>
+#include <ATen/cuda/CUDABlas.h>
+#include <ATen/native/ConvUtils.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/cuda/im2col.cuh>
 
-std::tuple<Tensor&, Tensor&, Tensor&> slow_conv2d_backward_out_cuda(const Tensor& grad_output,
-    const Tensor& self,
-    const Tensor& weight,
+namespace at { namespace native {
+namespace {
+
+void slow_conv2d_shape_check(
+    const Tensor& input, const Tensor& grad_output,
+    const Tensor& weight, const Tensor& bias,
+    int64_t kH, int64_t kW,
+    int64_t dH, int64_t dW,
+    int64_t padH, int64_t padW,
+    bool weight_nullable) {
+  TORCH_CHECK(kW > 0 && kH > 0,
+              "kernel size should be greater than zero, but got kH: ", kH, " kW: ", kW);
+  TORCH_CHECK(dW > 0 && dH > 0,
+              "stride should be greater than zero, but got dH: ", dH, " dW: ", dW);
+
+  TORCH_CHECK(weight_nullable || weight.defined(),
+              "weight tensor is expected to be non-nullable");
+  TORCH_CHECK(!weight.defined() ||
+              ((weight.numel() > 0) && (weight.dim() == 2 || weight.dim() == 4)),
+              "non-empty 2D or 4D weight tensor expected, but got: ", weight.sizes());
+  TORCH_CHECK(!bias.defined() || (bias.dim() == 1 && bias.sizes()[0] == weight.sizes()[0]),
+              "Expected bias to have shape [", weight.sizes()[0], "] but got ", bias.sizes());
+
+  const auto in_sizes = input.sizes();
+  int ndim = in_sizes.size();
+  int dimf = 0;
+  int dimh = 1;
+  int dimw = 2;
+
+  if (ndim == 4) {
+    dimf++;
+    dimh++;
+    dimw++;
+  }
+
+  // Allow for empty batch size but not other dimensions
+  const bool valid_empty = c10::multiply_integers(in_sizes.slice(1)) != 0;
```

Comment:
This is not right if ndim is 3 (the you need in_sizes.numel() > 0)? Btw, do we still allow 3d inputs, or is this code path never exercised?

Reply:
You're right, `_convolution` enforce the shape check before this is ever called. Should I just remove all the `dim=3` handling code?

Reply:
Yes, please do